design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,205 @@
+--------------
+```
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
+
+name: my-postgres-jdbc
+connector: io.confluent.connect.jdbc.SourceConnector
+topic-prefix: test-postgres-jdbc
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
+page-table | test-postgres-jdbc-page-table | id VARCHAR, page VARCHAR, url VARCHAR
+```
+
+Then the user can create streams or tables from the underlying topics:
+```sql
+ksql> CREATE TABLE users_original WITH (kafka_topic = 'test-postgres-jdbc-user-table', ...);
+
+--------------
+STREAM CREATED
+--------------
+```
+
+At this point, assuming the `pageviews_original` stream is already created, we can join the newly
+created `users_original` table with the `pageviews_original` stream. Finally, the user can drop and 
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc TERMINATE;
+```
+
+**Sample Egress**
+
+Data egress would work in a similar fashion:
+```sql
+ksql> CREATE EXTERNAL SINK my-elastic-sink FROM my-table1 WITH (...);
+
+ksql> SHOW EXTERNAL SINKS;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-elastic-sink  | io.confluent.connect.elasticsearch.ElasticsearchSinkConnector | elastic-sink
+
+ksql> DESCRIBE my-elastic-sink;
+
+name: my-elastic-sink
+connector: io.confluent.connect.elasticsearch.ElasticsearchSinkConnector
+topic-prefix: elstic-sink
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 1    | RUNNING
+
+name           | topic
+--------------------------------------------
+elasticsearch1 | elastic-sink-elasticsearch1 
+
+ksql> DROP EXTERNAL SINK my-elastic-sink TERMINATE;
+```
+
+### Advanced Topics
+
+- **Error Handling**: it is difficult to understand whether a given connector is or is not working,
+  and further difficult to debug why it is not working. This is out of scope for this KLIP and
+  should be addressed in connect and then exposed externally. There is a class of synchronous errors 
```

Comment:
Connect does expose status / errors etc though its API and we should be covering it as part of the MVP / KLIP.

There may be feedback we can give when making this integration and improvements we can make, but we should at least make initial use of what is there.

Reply:
Definitely - see the `DESCRIBE <connector>` section: that will list out whether or not a certain connector is functioning or not, which is the basic MVP requirement. 

I'm not sure that we should expose more than this - KSQL's strength isn't in helping you visualize error information and debug connect. If we went any further then this it would likely cause a large maintenance burden whenever connect changes.

Reply:
I'm not we can say KSQL shouldn't be exposing more here. We are positioning this as an integrated solution, so we should be enabling users to setup these connectors purely through the KSQL APIs, not forcing them to use the Connect API to work out what is going on.  We should be surfacing as much information as possible from Connect to allow users to get things working.